### PR TITLE
Do not include rvalue constructor in Algebraic if it already has a copy constructor

### PR DESCRIPTION
**Pull request by RazvanN7**
This currently blocks: https://github.com/dlang/dmd/pull/13687

If this gets merged could a new tag be pushed so that buildkite gets unblocked?

Thanks!

### CODE REVIEW

File: source/mir/algebraic.d [hunk 1]
```diff
@@ -1080,7 +1080,7 @@ struct Algebraic(T__...)
 
     /// Construct an algebraic type from its subset.
     this(RhsTypes...)(Algebraic!RhsTypes rhs)
-        if (allSatisfy!(Contains!AllowedTypes, Algebraic!RhsTypes.AllowedTypes))
+        if (!hasElaborateCopyConstructor!(Algebraic!T__)  && allSatisfy!(Contains!AllowedTypes, Algebraic!RhsTypes.AllowedTypes))
```
Comment (9il) [inline source/mir/algebraic.d:1083]:
```suggestion
        if (!(hasElaborateCopyConstructor!(Algebraic!T__) && is(Algebraic!RhsTypes == typeof(this)))  && allSatisfy!(Contains!AllowedTypes, Algebraic!RhsTypes.AllowedTypes))
```

**Comment by 9il** [comment]
Can copy constructor accept rvalue?

**Comment by RazvanN7** [comment]
Short answer: no. Long answer: if -preview=rvaluerefparam is used, technically, yes, but only when called explicitly, however, -preview=rvaluerefparam is not implemented properly, so I would advise against using it. On the flip side, you can try it if you want and see how that works out for you.

**Comment by 9il** [comment]
That is a breaking change we unlikely want to introduce. Why can't we preserve the current API?

**Comment by RazvanN7** [comment]
Because if a struct defines both an rvalue constructor and a copy constructor you end up with an infinite loop when trying to do overload resolution. For example, if you have `T a = b`, the compiler will rewrite it to `a.__ctor(b)`. Since both the rvalue constructor and copy constructor are in the same overload set, the rvalue constructor is tried; but since the rvalue constructor receives the argument by value it will have to be rewritten to `a.__ctor(tmp.__ctor(b))` and so on so forth. When I implemented copy construction I proposed that the copy constructor is in a different overload set than normal constructors, that would have permitted having both copy and rvalue constructors, but Walter rejected the proposal on the basis that it complicates the implementation. Hence, rvalue and copy constructors cannot be both defined at the same time.

Now with templated functions, the story is different. The compiler cannot know if a particular template could result in an rvalue constructor, therefore up until this point it did not do any checks on it, but that may lead to infinite loop in the compiler as observed in: https://issues.dlang.org/show_bug.cgi?id=21613 . So when doing overload resolution over constructors, if an rvalue constructor that is the result of a template instance is identified, the compiler will error because the potential of breakage is there.

Since we've gone down that path, I don't see any way in which the behavior in mir.core.algebraic is allowed so it's better to just swallow the pill and merge this.

I hope this answers your question.

**Comment by 9il** [comment]
> Since both the rvalue constructor and copy constructor are in the same overload set, the rvalue constructor is tried

Why does the rvalue constructor has priority? I expected the opposite behaviour.

**Comment by RazvanN7** [comment]
It does not have priority, but since they are both defined they need to be checked when overload resolution occurs.

Currently if you have:

```d
void fun(int a) {}
void fun(ref int a) {}
```

When called with `fun(d)`, the compiler checks but fun functions and gets perfect matches for both, then it checks which one is more specialized and sees that `fun(ref int)` is more specialized and picks that one. The same algorithm is applied for constructors, but you enter an infinite loop when checking the rvalue constructor.

**Comment by jmh530** [comment]
@RazvanN7 So you're saying that the error occurs when it tries the move constructor, not that the move constructor gets used per se. Because I would have thought that the copy constructor is the more specialized one since it takes arguments by ref. 

It sounds like this is a perfect time to go back to Walter about how copy constructors are implemented.

**Comment by 9il** [comment]
I can't get why it can't be resolved in the compiler. I am not a compiler expert. That sounds like a compiler implementation limitation rather than a language limitation.

**Comment by RazvanN7** [comment]
As mentioned before, when I proposed the copy construction implementation this was not a limitation, but Walter insisted on it saying that [it complicates the implementation and that the rules will be hard to understand](https://github.com/dlang/dmd/pull/8688#discussion_r250134934).

Indeed, we could have allowed this scenario, however, as stated in the comment I linked it adds additional complexity to the implementation. We have been on this path (not allowing rvalue constructors alongside copy constructors) from the beginning and nobody complained about it. However, you could still have templated constructors that for some instances might result in rvalue constructors. In this case you would have been faced with a compiler infinite loop so it's unlikely that this change is a breaking change.

So I don't really get the opposition for this change. If you want to convince Walter to accept rvalue constructors alongside copy constructors, be my guest, I've done my part. However, that should not block this PR being merged since this currently stops a patch that would prevent the compiler from crashing in certain situations.

**Comment by RazvanN7** [comment]
@9il Thank you!

**Pull request merged**